In [1]:
import psycopg2
import pandas as pd

import sys
sys.path.append("../")

import src.soporte_bbdd as sb

# Limpieza de peliculas

In [2]:
df_peliculas = pd.read_csv("../datos/output/peliculas.csv")

In [3]:
df_peliculas.head()

Tipo                         Titulo   Año  Mes         Id generos
0  Movie  Las Travesuras de Super Chido  1990   10  tt0320656  Action
1  Movie            Velai Kidaichiruchu  1990    8  tt0318803  Action
2  Movie                    Sor Batalla  1990    8  tt0320503  Action
3  Movie            Azaad Desh Ke Gulam  1990    4  tt0330019  Action
4  Movie               Suo ming fei dao  1991   11  tt0348147  Action

In [4]:
sb.exploracion_datos(df_peliculas)

El número de filas es 1453 y el número de columnas es 6

----------

En este conjunto de datos tenemos 0 valores duplicados

----------

Los columnas con valores nulos y sus porcentajes son: 


Series([], dtype: float64)


----------

Las principales estadísticas de las variables númericas son:


count         mean         std     min     25%     50%     75%     max
Año  1453.0  2015.571232  209.771911  1990.0  2003.0  2012.0  2018.0  9999.0
Mes  1453.0     5.773572    3.650211     0.0     3.0     5.0     9.0    12.0


----------

Las principales estadísticas de las variables categóricas son:


count unique         top freq
Tipo     1453      4       Movie  776
Titulo   1453   1444        Hush    2
Id       1453   1453  tt27262238    1
generos  1453      8      Horror  405


----------

Las características principales del dataframe son:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1453 entries, 0 to 1452
Data columns (total 6 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   Tipo     1453 non-null   object
 1   Titulo   1453 non-null   object
 2   Año      1453 non-null   int64 
 3   Mes      1453 non-null   int64 
 4   Id       1453 non-null   object
 5   generos  1453 non-null   object
dtypes: int64(2), object(4)
memory usage: 68.2+ KB


None

In [5]:
condicion = df_peliculas["Año"] <= 2024
df_peliculas = df_peliculas[condicion]

# Limpieza de pelicula y artista

In [6]:
df_pelicula_artista = pd.read_csv("../datos/output/pelicula_artista.csv")
df_pelicula_artista.head()

Movie ID       Actriz_Actor
0  tt0110660  Tony Ka Fai Leung
1  tt0110363       Luke Edwards
2  tt0110397    Jennifer Barker
3  tt0111045        Miklós Déri
4  tt0111653           Joe Bays

In [7]:
sb.exploracion_datos(df_pelicula_artista)

El número de filas es 14529 y el número de columnas es 2

----------

En este conjunto de datos tenemos 3298 valores duplicados

----------

Los columnas con valores nulos y sus porcentajes son: 


Series([], dtype: float64)


----------

Las principales estadísticas de las variables númericas son:


count unique        top  freq
Movie ID      14529   1452  tt0929793    20
Actriz_Actor  14529   9909  Sin datos  4094


----------

Las principales estadísticas de las variables categóricas son:


count unique        top  freq
Movie ID      14529   1452  tt0929793    20
Actriz_Actor  14529   9909  Sin datos  4094


----------

Las características principales del dataframe son:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14529 entries, 0 to 14528
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Movie ID      14529 non-null  object
 1   Actriz_Actor  14529 non-null  object
dtypes: object(2)
memory usage: 227.1+ KB


None

In [8]:
df_pelicula_artista[df_pelicula_artista.duplicated()]

Movie ID   Actriz_Actor
1153    tt0929793  Michele Cantù
1607   tt12141548      Sin datos
1613   tt13820914      Sin datos
1622   tt12572368      Sin datos
1647    tt2043811      Sin datos
...           ...            ...
14523  tt27236215      Sin datos
14524  tt27251096      Sin datos
14525  tt27250288      Sin datos
14526  tt27250168      Sin datos
14528  tt10774576      Sin datos

[3298 rows x 2 columns]

In [9]:
df_pelicula_artista.drop_duplicates()

Movie ID       Actriz_Actor
0       tt0110660  Tony Ka Fai Leung
1       tt0110363       Luke Edwards
2       tt0110397    Jennifer Barker
3       tt0111045        Miklós Déri
4       tt0111653           Joe Bays
...           ...                ...
14504  tt10709060   Kendra Kuhnhofer
14514  tt14142640            Prakash
14519  tt18688348    Oleg Primogenov
14522  tt27241513   Georgiy Topolaga
14527  tt27262238    Jacob Schindler

[11231 rows x 2 columns]

# Limpieza del detalle de las peliculas

In [10]:
df_detalles_peliculas = pd.read_csv("../datos/output/detalles_peliculas.csv")
df_detalles_peliculas.head()

titulo  cali_imdb             direc_imbd  \
0  Las Travesuras de Super Chido        7.7         Alejandro Todd   
1            Velai Kidaichiruchu        5.2                P. Vasu   
2                    Sor Batalla        6.6  Jesús Fragoso Montoya   
3            Azaad Desh Ke Gulam        5.3    S.A. Chandrashekhar   
4               Suo ming fei dao        NaN            Desconocido   

                       guion_imbd  \
0      Óscar FentanesJuan Garrido   
1        SathyarajGautamiAnandraj   
2                Fernando Galiana   
3  RekhaRishi KapoorJackie Shroff   
4                     Desconocido   

                                          argum_imbd   durac_imbd id_pelicula  
0                                        Desconocido     1h 35min   tt0320656  
1  Baasha, a gangster, controls a whole village w...  Desconocido   tt0318803  
2  As punishment for participating in a contest T...     1h 34min   tt0320503  
3  Bharti, a law student from a rich family, live...     2h 33min   tt0330019  
4                                        Desconocido     1h 24min   tt0348147

In [11]:
sb.exploracion_datos(df_detalles_peliculas)

El número de filas es 1453 y el número de columnas es 7

----------

En este conjunto de datos tenemos 0 valores duplicados

----------

Los columnas con valores nulos y sus porcentajes son: 


cali_imdb    33.860977
dtype: float64


----------

Las principales estadísticas de las variables númericas son:


count      mean       std  min  25%  50%  75%  max
cali_imdb  961.0  5.912071  1.426717  1.3  5.0  6.0  6.9  9.8


----------

Las principales estadísticas de las variables categóricas son:


count unique          top freq
titulo       1453   1444        Guilt    2
direc_imbd   1453    858  Desconocido  576
guion_imbd   1453    906  Desconocido  539
argum_imbd   1453    682  Desconocido  772
durac_imbd   1453    231  Desconocido  231
id_pelicula  1453   1453   tt27215062    1


----------

Las características principales del dataframe son:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1453 entries, 0 to 1452
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   titulo       1453 non-null   object 
 1   cali_imdb    961 non-null    float64
 2   direc_imbd   1453 non-null   object 
 3   guion_imbd   1453 non-null   object 
 4   argum_imbd   1453 non-null   object 
 5   durac_imbd   1453 non-null   object 
 6   id_pelicula  1453 non-null   object 
dtypes: float64(1), object(6)
memory usage: 79.6+ KB


None

In [12]:
df_detalles_peliculas.dropna(inplace=True)
df_detalles_peliculas.reset_index(inplace=True, drop=True)

# Limpieza de los artistas

In [13]:
df_detalles_artistas = pd.read_csv("../datos/output/detalles_artistas.csv")
df_detalles_artistas.head()

actr  año_nac  \
0          A. Gunaseelan     9999   
1  A. Michelle Harleston     9999   
2          A.C. Peterson     9999   
3     A.V.S. Subramanyam     1957   
4          Øyvind Øvrebø     9999   

                                        conocido_por  \
0                                          'Mandela'   
1  'Murder by Numbers', 'The Mark of a Killer', '...   
2  'Narc', 'Shanghai Noon', 'Shooter', 'The Last ...   
3  'Roommates', 'Uncle', 'Ori Nee Prema Bangaramk...   
4  'Tingen, Edderkoppen', 'Demenskoret', 'Kristia...   

                                        que_hace           premios  
0  Casting Department, Actress, Casting Director      1 nomination  
1                                          Actor  no tiene premios  
2                                          Actor  no tiene premios  
3                        Actor, Director, Writer  no tiene premios  
4                       Actress, Additional Crew  no tiene premios

In [14]:
sb.exploracion_datos(df_detalles_artistas)

El número de filas es 9907 y el número de columnas es 5

----------

En este conjunto de datos tenemos 0 valores duplicados

----------

Los columnas con valores nulos y sus porcentajes son: 


Series([], dtype: float64)


----------

Las principales estadísticas de las variables númericas son:


count         mean          std     min     25%     50%     75%  \
año_nac  9907.0  6685.205309  3955.479167  1860.0  1971.0  9999.0  9999.0   

            max  
año_nac  9999.0


----------

Las principales estadísticas de las variables categóricas son:


count unique                                      top  freq
actr          9907   9907                         Zuzana Mráziková     1
conocido_por  9907   9301  'No hay datos sobre porqué es conocido'   170
que_hace      9907   1076                                    Actor  2355
premios       9907    872                         no tiene premios  6075


----------

Las características principales del dataframe son:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9907 entries, 0 to 9906
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   actr          9907 non-null   object
 1   año_nac       9907 non-null   int64 
 2   conocido_por  9907 non-null   object
 3   que_hace      9907 non-null   object
 4   premios       9907 non-null   object
dtypes: int64(1), object(4)
memory usage: 387.1+ KB


None

In [15]:
condicion = df_detalles_artistas["año_nac"] > 2024
df_nulos = df_detalles_artistas[condicion]

In [16]:
df_nulos.head()

actr  año_nac  \
0          A. Gunaseelan     9999   
1  A. Michelle Harleston     9999   
2          A.C. Peterson     9999   
4          Øyvind Øvrebø     9999   
5             Óscar Abad     9999   

                                        conocido_por  \
0                                          'Mandela'   
1  'Murder by Numbers', 'The Mark of a Killer', '...   
2  'Narc', 'Shanghai Noon', 'Shooter', 'The Last ...   
4  'Tingen, Edderkoppen', 'Demenskoret', 'Kristia...   
5  'El prado de las estrellas', 'The Little Thing...   

                                        que_hace           premios  
0  Casting Department, Actress, Casting Director      1 nomination  
1                                          Actor  no tiene premios  
2                                          Actor  no tiene premios  
4                       Actress, Additional Crew  no tiene premios  
5                                        Actress  no tiene premios

- No quitamos nulos (nacidos en 9999) de este conjunto de datos porque supone una gran perdida de información y siempre que hagamos las querys en SQL podemos tener en cuenta condiciones para no obtener este tipo de datos.

# Creación de la bbdd

Tablas: Peliculas, Artistas, Reparto, Calificaciones

`Peliculas` tiene relacion con `Reparto` y `Calificaciones`

`Reparto` tiene relación con `Artistas`

In [17]:
try:
    connection = psycopg2.connect(
        dbname="postgres",
        user="postgres",
        password="admin",
        host="localhost",
    )
    connection.autocommit = True
    nombre_bbdd = "prueba_tecnica"

    # Crear un cursor para ejecutar comandos SQL
    cursor = connection.cursor()

    cursor.execute("SELECT 1 FROM pg_database WHERE datname = %s", (nombre_bbdd,))
    
    # Almacenamos en una variable el resultado del fetchone. Si existe tendrá una fila sino será None
    bbdd_existe = cursor.fetchone()
    
    # Si bbdd_existe es None, creamos la base de datos
    if not bbdd_existe:
        cursor.execute("CREATE DATABASE prueba_tecnica;")
        print(f"Base de datos {nombre_bbdd} creada con éxito")
    else:
        print(f"La base de datos ya existe")

    connection = psycopg2.connect(
        dbname=nombre_bbdd,
        user="postgres",
        password="admin",
        host="localhost",
    )

    # Definir la estructura de las tablas
    query_create_peliculas = """
    CREATE TABLE IF NOT EXISTS peliculas (
        id_peli INT PRIMARY KEY,
        titulo VARCHAR(300),
        tipo VARCHAR(300),
        anio NUMERIC,
        mes NUMERIC,
        generos VARCHAR(300)
    )"""
    cursor.execute(query_create_peliculas)

    query_create_artistas = """
    CREATE TABLE IF NOT EXISTS artistas (
        id_artista SERIAL PRIMARY KEY,
        nombre VARCHAR(300),
        anio NUMERIC,
        conocido_por VARCHAR(300),
        que_hace VARCHAR(300),
        premios VARCHAR(300)
    )"""
    cursor.execute(query_create_artistas)

    query_create_reparto = """
    CREATE TABLE IF NOT EXISTS reparto (
        id_peli INT PRIMARY KEY REFERENCES peliculas(id_peli),
        id_artista INT PRIMARY KEY REFERENCES artistas(id_artista)
    )"""
    cursor.execute(query_create_reparto)

    query_create_calificaciones = """
    CREATE TABLE IF NOT EXISTS calificaciones (
        id_peli INT PRIMARY KEY REFERENCES peliculas(id_peli),
        titulo VARCHAR(300),
        puntuacion VARCHAR(300),
        director VARCHAR(300),
        guionistas VARCHAR(300),
        argumento VARCHAR(600),
        duracion VARCHAR(300)
    )"""
    cursor.execute(query_create_calificaciones)
    
except Exception as e:
    connection.rollback()
    print("Error al conectar a la base de datos:", e)
    raise


La base de datos ya existe


In [19]:
df_peliculas.reindex(columns=["Id", "Titulo", "Tipo", "Año", "Mes", "generos"])

Id                                             Titulo      Tipo  \
0      tt0320656                      Las Travesuras de Super Chido     Movie   
1      tt0318803                                Velai Kidaichiruchu     Movie   
2      tt0320503                                        Sor Batalla     Movie   
3      tt0330019                                Azaad Desh Ke Gulam     Movie   
4      tt0348147                                   Suo ming fei dao     Movie   
...          ...                                                ...       ...   
1448  tt27236215                                              Sight     Short   
1449  tt27251096  Der Wien-Krimi: Blind ermittelt - Tod an der D...  TV Movie   
1450  tt27250288                           Marne Ke Dar Se Mere Dil     Movie   
1451  tt27250168                                           Red Wine     Short   
1452  tt27262238                                      Magna Lacerta     Movie   

       Año  Mes  generos  
0     1990   10   Action  
1     1990    8   Action  
2     1990    8   Action  
3     1990    4   Action  
4     1991   11   Action  
...    ...  ...      ...  
1448  2023    1  Mystery  
1449  2023    4  Mystery  
1450  2023    3  Mystery  
1451  2023   11  Mystery  
1452  2023    3  Mystery  

[1452 rows x 6 columns]

In [ ]:
df_detalles_peliculas = df_detalles_peliculas.reindex(columns=["id_pelicula", "titulo", "cali_imdb", "direc_imdb", "guion_imdb", "argum_imdb", "durac_imdb"])
df_detalles_peliculas.head()

- Antes de realizar la inserción con la función que tenemos debemos renombrar las columnas y ponerlas en el orden correspondiente

In [ ]:
sb.insertar_datos(cursor, df_peliculas, "peliculas")

sb.insertar_datos(cursor, df_detalles_artistas, "artistas")

sb.insertar_datos(cursor, df_pelicula_artista, "reparto")

sb.insertar_datos(cursor, df_detalles_peliculas, "calificaciones")

cursor.close()
connection.close()

In [ ]:
query_create_calificaciones = """
    SELECT generos
    FROM peliculas
"""

- Para realizar las consultas correspondientes debemos realizar joins de las tablas para poder obtener los resultados que se piden.